### PR TITLE
Fix bug: Property does not work when setting jms pool disabled

### PR DESCRIPTION
**Pull request by moarychan**
# Description

Fixes https://github.com/Azure/azure-sdk-for-java/issues/43279

Property does not work when setting spring.jms.servicebus.pool.enabled=false using property source.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+import static org.springframework.beans.factory.support.BeanDefinitionBuilder.genericBeanDefinition;
+
+/**
+ * Configuration for {@link ConnectionFactory}.
+ *
+ * @since 5.19.0
+ */
+@Import(AzureJmsConnectionFactoryConfiguration.Registrar.class)
+public class AzureJmsConnectionFactoryConfiguration {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java:39]: Can this be non-public?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                registryJmsPoolConnectionFactory(registry);
+                return;
+            }
+
+            if (isCacheConnectionFactoryClassPresent() && (!cacheEnabledResult.isBound() || cacheEnabledResult.orElseGet(() -> false))) {
+                registryJmsCacheConnectionFactory(registry);
+                return;
+            }
+
+            if (cacheEnabledResult.isBound() && !cacheEnabledResult.orElseGet(() -> false)) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java:75]: Do we need this condition here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsAutoConfiguration.java [hunk 3]
```diff
@@ -47,11 +46,12 @@
     AzureServiceBusResourceManagerAutoConfiguration.class })
 @ConditionalOnProperty(value = "spring.jms.servicebus.enabled", matchIfMissing = true)
 @ConditionalOnClass({ ConnectionFactory.class, JmsConnectionFactory.class, JmsTemplate.class })
-@EnableConfigurationProperties({ JmsProperties.class })
+@EnableConfigurationProperties
 @Import({
-    ServiceBusJmsConnectionFactoryConfiguration.class,
+    AzureJmsConnectionFactoryConfiguration.class,
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsAutoConfiguration.java:51]: Why don't we still call it `ServiceBusJmsConnectionFactoryConfiguration`

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+        @Override
+        public void registerBeanDefinitions(AnnotationMetadata importingClassMetadata,
+                                            BeanDefinitionRegistry registry) {
+            BindResult<Boolean> poolEnabledResult = Binder.get(environment).bind("spring.jms.servicebus.pool.enabled", Boolean.class);
+            BindResult<Boolean> cacheEnabledResult = Binder.get(environment).bind("spring.jms.cache.enabled", Boolean.class);
+
+            if (isPoolConnectionFactoryClassPresent()
+                && ((!cacheEnabledResult.isBound() && !poolEnabledResult.isBound())) || poolEnabledResult.orElseGet(() -> false)) {
+                registryJmsPoolConnectionFactory(registry);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java:66]:
```suggestion
                registerJmsPoolConnectionFactory(registry);
```

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                    ObjectProvider<AzureServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers = beanFactory.getBeanProvider(AzureServiceBusJmsConnectionFactoryCustomizer.class);
+                    return createJmsConnectionFactory(properties, factoryCustomizers);
+                });
+            registry.registerBeanDefinition(JMS_CONNECTION_FACTORY_BEAN_NAME, definitionBuilder.getBeanDefinition());
+        }
+
+        private void registryJmsCacheConnectionFactory(BeanDefinitionRegistry registry) {
+            BeanDefinitionBuilder definitionBuilder = genericBeanDefinition(CachingConnectionFactory.class,
+                () -> {
+                    JmsProperties jmsProperties = beanFactory.getBean(JmsProperties.class);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfiguration.java:102]: Seems like some code duplication

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfigurationTests.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import org.springframework.boot.autoconfigure.jms.JmsAutoConfiguration;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.jms.connection.CachingConnectionFactory;
+
+import static com.azure.spring.cloud.autoconfigure.implementation.util.TestServiceBusUtils.CONNECTION_STRING_FORMAT;
+import static org.assertj.core.api.Assertions.assertThat;
+
+class AzureJmsConnectionFactoryConfigurationTests {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/implementation/jms/AzureJmsConnectionFactoryConfigurationTests.java:21]: Please also rename this one.

**Review by saragluna — CHANGES_REQUESTED** [review]

**Review by saragluna — APPROVED** [review]

File: sdk/spring/CHANGELOG.md [hunk 7]
```diff
@@ -14,6 +14,7 @@ This section includes changes in `spring-cloud-azure-autoconfigure` module.
 - Fix error: Service Bus connection string is still used when it's configured to empty string. [#42880](https://github.com/Azure/azure-sdk-for-java/issues/42880).
 - Fix warn logs: Mark the bean post processor related beans as internal to avoid the log 'Bean xxx of type xxx is not eligible for getting processed by all BeanPostProcessors'. [#38631](https://github.com/Azure/azure-sdk-for-java/issues/38631).
 - Fix missing bean `springTokenCredentialProviderContextProvider`: Azure Identity Extensions and Spring Cloud Stream Binder cannot work together. [#43147](https://github.com/Azure/azure-sdk-for-java/issues/43147).
+- Fix bug: Property does not work when setting `spring.jms.servicebus.pool.enabled=false` using property source. [#43279](https://github.com/Azure/azure-sdk-for-java/issues/43279).
```
[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/CHANGELOG.md]: Added the related annotation to indicate the scenario.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 8]
```diff
@@ -7,77 +7,135 @@
...
 
-    }
+        private static boolean isCacheConnectionFactoryClassPresent() {
+            return ClassUtils.isPresent("org.springframework.jms.connection.CachingConnectionFactory", null);
+        }
+
+        private static boolean isPoolConnectionFactoryClassPresent() {
+            return ClassUtils.isPresent("org.messaginghub.pooled.jms.JmsPoolConnectionFactory", null)
+                && ClassUtils.isPresent("org.apache.commons.pool2.PooledObject", null);
+        }
```
[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java:85]: Added UTs and ITs.

**Comment by Saljack** [comment]
Honestly, this is the most terrible idea of how to fix it. It just complicates everything and looks like you try to fix something broken by something more broken. None of you will understand what you are trying to achieve with this code in a few months. You should be aware that this is a really bad way how to write any autoconfiguration. I am really surprised that it passed through your code review. Isn't it possible to do it simple stupid? I would remove the whole `AzureServiceBusJmsPropertiesEnvironmentPostProcessor` and use only `@ConditionalOnXXX`
```java
    @Configuration(proxyBeanMethods = false)
    @ConditionalOnMissingBean(JmsPoolConnectionFactory.class)
    class SimpleConnectionFactoryConfiguration {
      // ...
    }

    @Configuration(proxyBeanMethods = false)
    @ConditionalOnClass({ JmsPoolConnectionFactory.class, PooledObject.class })
    class PooledConnectionFactoryConfiguration {

        @Bean(destroyMethod = "stop")
        @ConditionalOnProperty(prefix = "spring.jms.servicebus.pool", name = "enabled", havingValue = "true", matchIfMissing = true)
        JmsPoolConnectionFactory jmsPoolConnectionFactory(AzureServiceBusJmsProperties properties,
                                                          ObjectProvider<AzureServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
            ServiceBusJmsConnectionFactory factory = createJmsConnectionFactory(properties, factoryCustomizers);
            return new JmsPoolConnectionFactoryFactory(properties.getPool())
                .createPooledConnectionFactory(factory);
        }
    }
```
Another way is introducing a new configuration property `spring.jms.servicebus.connection.factory` with three values: `POOL`, `CACHE` and `SIMPLE`.

**Comment by saragluna** [comment]
I think I can explain why we chose to implement this way, but we do want to improve this if there's a better way (sorry for any inconvenience):

Before we used https://github.com/Azure/azure-sdk-for-java/blob/b44403740b282f5c1abbf0c01ccd429aa909a0bb/sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/implementation/jms/ServiceBusJmsConnectionFactoryConfiguration.java#L35-L81

so the value matrix would be 


| property/value | spring.jms.servicebus.pool | spring.jms.cache | registered bean type |
| :-: | :- | :- | :-|
| | empty | empty | CachingConnectionFactory |
| | empty | true  | CachingConnectionFactory |
| | empty | false | ServiceBusJmsConnectionFactory |
| | true  | empty | JmsPoolConnectionFactory |
| | true  | true  | JmsPoolConnectionFactory |
| | true  | false | JmsPoolConnectionFactory |
| | false | empty | CachingConnectionFactory |
| | false | true  | CachingConnectionFactory |
| | false | false | ServiceBusJmsConnectionFactory |


But we only want to change the default behavior when there's no `spring.jms.servicebus.pool` or `spring.jms.cache` provided. If we change the code to 

```java
@Configuration(proxyBeanMethods = false)
@ConditionalOnProperty(prefix = "spring.jms.servicebus.pool", name = "enabled", havingValue = "false",
    matchIfMissing = false)
static class SimpleConnectionFactoryConfiguration {

    @Bean
    @ConditionalOnProperty(prefix = "spring.jms.cache", name = "enabled", havingValue = "false")
    ServiceBusJmsConnectionFactory jmsConnectionFactory(AzureServiceBusJmsProperties properties,
                                                        ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
        return createJmsConnectionFactory(properties, factoryCustomizers);
    }

    @Configuration(proxyBeanMethods = false)
    @ConditionalOnClass(CachingConnectionFactory.class)
    @ConditionalOnProperty(prefix = "spring.jms.cache", name = "enabled", havingValue = "true",
        matchIfMissing = true)
    static class CachingConnectionFactoryConfiguration {

        @Bean
        CachingConnectionFactory jmsConnectionFactory(JmsProperties jmsProperties,
                                                      AzureServiceBusJmsProperties properties,
                                                      ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
            ServiceBusJmsConnectionFactory factory = createJmsConnectionFactory(properties, factoryCustomizers);
            CachingConnectionFactory connectionFactory = new CachingConnectionFactory(factory);
            JmsProperties.Cache cacheProperties = jmsProperties.getCache();
            connectionFactory.setCacheConsumers(cacheProperties.isConsumers());
            connectionFactory.setCacheProducers(cacheProperties.isProducers());
            connectionFactory.setSessionCacheSize(cacheProperties.getSessionCacheSize());
            return connectionFactory;
        }
    }

}

@Configuration(proxyBeanMethods = false)
@ConditionalOnClass({ JmsPoolConnectionFactory.class, PooledObject.class })
static class PooledConnectionFactoryConfiguration {

    @Bean(destroyMethod = "stop")
    @ConditionalOnProperty(prefix = "spring.jms.servicebus.pool", name = "enabled", havingValue = "true", matchIfMissing = true)
    JmsPoolConnectionFactory jmsPoolConnectionFactory(AzureServiceBusJmsProperties properties,
                                                      ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
        ServiceBusJmsConnectionFactory factory = createJmsConnectionFactory(properties, factoryCustomizers);
        return new JmsPoolConnectionFactoryFactory(properties.getPool())
            .createPooledConnectionFactory(factory);
    }
}
```

The matrix would be

| property/value | spring.jms.servicebus.pool | spring.jms.cache | registered bean type |
| :-: | :- | :- | :-|
| | empty | empty | JmsPoolConnectionFactory |
| | empty | true  | JmsPoolConnectionFactory |
| | empty | false | JmsPoolConnectionFactory |
| | true  | empty | JmsPoolConnectionFactory |
| | true  | true  | JmsPoolConnectionFactory |
| | true  | false | JmsPoolConnectionFactory |
| | false | empty | CachingConnectionFactory |
| | false | true  | CachingConnectionFactory |
| | false | false | ServiceBusJmsConnectionFactory |

**Comment by saragluna** [comment]
Yes, we could do that. I admit that the registrar implementation is not as straightforward as auto-configuration, but could you please share what's the current problem function-wise?

**Comment by Saljack** [comment]
It is not a function-wise problem (I have not tested it yet). But if someone checks this code, it is hard to understand and it is not common. It can happen because combinations of JMS factory properties are not clear at first sight. This combination can surprise new users (I understand that you want to be backward compatible):
| spring.jms.servicebus.pool |	spring.jms.cache | registered bean type |
| ----------------------------- | ------------------- | ---------------------- |
| empty                                 | false                      | ServiceBusJmsConnectionFactory |

 I just think this code is ugly, hard to maintain and extend. Mainly if there is any better and easier option with a custom `@Conditial`. But it is just my opinion and it is your code and your decision.

**Pull request merged**